go/dpagg/mean.go

```diff
@@ -213,6 +216,102 @@ func (bm *BoundedMeanFloat64) Result() float64 {
+	// a fixed number of splits numSplits for countAlpha where it takes the
+	// range of values strictly from 0 to alpha.
+	increment := alpha / float64(numSplits)
+	curCountAlpha := increment
+	bestTightness := math.Inf(1)
+	var bestConfInt noise.ConfidenceInterval
+	for curCountAlpha < alpha {
+		confInt, err := bm.computeConfidenceIntervalSplit(alpha, curCountAlpha)
+		if err != nil {
+			return noise.ConfidenceInterval{}, err
+		}
+		tightness := confInt.UpperBound - confInt.LowerBound
+		if tightness < bestTightness {
+			bestTightness = tightness
+			bestConfInt = confInt
+		}
+		curCountAlpha += increment
+	}
+	return bestConfInt, nil
+}
+
+// As opposed to the former function, this function has countAlpha where you can set confidence
+// level for count confidence interval. It computes the coresponding alpha for sum confidence
+// interval to return a confidence interval with a confidence level of at least 1 - meanAlpha.
+//
+// Note Result() needs to be called before ComputeConfidenceInterval, otherwise this will return an error.
+func (bm *BoundedMeanFloat64) computeConfidenceIntervalSplit(meanAlpha, countAlpha float64) (noise.ConfidenceInterval, error) {
+	if !bm.resultReturned {
+		return noise.ConfidenceInterval{}, fmt.Errorf("You need to call Result() before calling ComputeConfidenceInterval()")
+	}
+	if err := checkSplit(meanAlpha, countAlpha); err != nil {
+		return noise.ConfidenceInterval{}, err
+	}
+
+	sumAlpha := (meanAlpha - countAlpha) / (1 - countAlpha)
+	sumConfInt, err := bm.normalizedSum.ComputeConfidenceInterval(sumAlpha)
+	if err != nil {
+		return noise.ConfidenceInterval{}, err
+	}
+	countConfInt, err := bm.count.ComputeConfidenceInterval(countAlpha)
+	if err != nil {
+		return noise.ConfidenceInterval{}, err
+	}
+
+	// The BoundedMean algorithms maximizes bounds in the denominator to 1, since count
+	// cannot be less than one and to avoid division by zero.
+	countConfInt.LowerBound = math.Max(countConfInt.LowerBound, 1)
+	countConfInt.UpperBound = math.Max(countConfInt.UpperBound, 1)
+
+	// Estimate the lower and upper bounds as minimum and maximum values for mean using lower and
+	// upper bounds of confidence intervals for sum and count.
+	var meanLowerBound, meanUpperBound float64
+	if sumConfInt.LowerBound >= 0 {
+		meanLowerBound = sumConfInt.LowerBound / countConfInt.UpperBound
+	} else {
+		meanLowerBound = sumConfInt.LowerBound / countConfInt.LowerBound
+	}
+	if sumConfInt.UpperBound >= 0 {
+		meanUpperBound = sumConfInt.UpperBound / countConfInt.LowerBound
+	} else {
+		meanUpperBound = sumConfInt.UpperBound / countConfInt.UpperBound
+	}
+	// We are using normalizedSum so we need to add midPoint to both bounds.
+	meanLowerBound, meanUpperBound = meanLowerBound+bm.midPoint, meanUpperBound+bm.midPoint
+
+	// Clamp mean bounds to lower and upper bounds.
+	meanLowerBound, err = ClampFloat64(meanLowerBound, bm.lower, bm.upper)
+	if err != nil {
+		return noise.ConfidenceInterval{}, err
```

Comment:
Maybe we could create a more meaningful error message here, something like `fmt.Errorf(Couldn't clamp lower bound for mean: %v, err)`. Otherwise it might be confusing what we are clamping and why.

(Same below)